### ISSUE TITLE
Improve UI/UX consistency between Shapes and Labels Polygon tools

### ISSUE THREAD

**Issue by psobolewskiPhD** [napari/napari#6489]
This came up in the docs working group in the context of a major update of the layer guides:
https://github.com/napari/docs/pull/125/files#diff-2f8f3abb006bb80c683093a552671ed1c5a1b79a6f1527195a1db862c44a6eba
So I think it may be worth looking at Polygons in the Shapes and Labels layer from the PoV of the user and the docs.
Importantly, @**MelissaWM** is also going to be updating the screenshots and movies, so it's important not to waste a ton of time if something isn't quite right.
Note: everything I'm referring to is referring to **0.5.0**

To recap: In Shapes layers we have two tools for drawing polygons. The original Polygon tool (triangle icon) and the Polygon lasso (lasso icon). The tools are next to each other in the toolbar.
![image](https://github.com/napari/napari/assets/76622105/43b83633-da1c-4af6-a8d1-5cf8454ba5c4)

They can both draw the same Shapes, but the Polygon tool requires clicks per vertex, while the lasso just requires moving the mouse and the vertexes are automatically placed. Both tools make closed shapes. Both tools end drawing in a similar but different fashion. In the original Polygon tool a *double click* ends and in the newer lasso a single click. At first this seems odd, but in the OG tool you are constantly clicking, so an extra click, the double click, means end, while in the lasso you click to place the origin and then you don't click, just drag, so a single click suffices to end and close the polygon.
Overall, I think these tools are pretty consistent and very complementary.

In the Labels layer there is a new tool to draw Polygons and it's called the Polygon tool, but has the lasso icon (to be released in 0.5.0). It's next to the paintbrush.
![image](https://github.com/napari/napari/assets/76622105/ca6e1fc0-328b-4a19-a069-cef1854c56ae)

Its behavior is very handy, because it makes filled labels easily. In fact, it's virtually identical to the *original Shapes Polygon tool* mentioned above: each click places a vertex of a closed shape. The visual behavior is basically the same. Note this is in contrast to the Polygon *lasso* tool of the Shapes layer, with which this tool shares the icon. The end behavior is similar to the Shapes tools, automatically closing the Shape and using double-click. However, there is a subtle difference, because the double click has to be within 20 pixels of the origin point of the polygon, otherwise it just sets another vertex:
https://github.com/napari/napari/blob/85c58b5990dfbbc41468ae16c25216e8a00f11a3/napari/components/overlays/labels_polygon.py#L38
Importantly this is screen pixels and does not appear to account for HiDPI, so on my new work monitor is actually annoying close to the origin when going quickly on larger annotations. Note that the rationale was largely due to slow double-click speed on linux resulting in un-wanted ending.

So, while we can document all of this--and have to some extent-- I think there is some inconsistency in these tools. First, the Labels Polygon tool icon is inconsistent Shapes tool icons, based on the behavior. The Labels Polygon tool behaves like the Shapes Polygon tool (triangle icon) not the Polygon lasso tool (lasso icon). 
- I think due to the similarity of behavior it would be best to have the icons consistent with each other. Does anyone have any thoughts there?

Likewise, I also find the close behavior of the Labels Polygon tool frustratingly inconsistent with Shapes Polygon tool, because it's very similar, yet different. As a creature of habit prefer the more common solution of double click to end a closed shape regardless of how close to origin. (Note: I can see having to double click close to the origin point to close a path, which could otherwise be open, but the Shape and Labels polygons are closed by default.) Here the next steps are less clear:
- Maybe the `completion_radius` should be exposed in the Preferences? Then people could tweak based on monitor Rez and how intricate their labels are to be?
- Maybe it should accept `None` or something (infinity!) then the double-click always closes, just like the Shapes tools?
- Should we re-consider what the default behavior should be?
- Maybe the *Shapes* tools should behave like the Labels tool does? But I'm loath to change something people are likely using and used to (plus my survey of tools I have access to suggests it's the more common behavior.)
Personally I think all 3 Polygon tools should have the same close behavior by default for consistency. And if there is a completion_radius setting it should affect all of them.

**Comment by jni** [comment]
> I think due to the similarity of behavior it would be best to have the icons consistent with each other. Does anyone have any thoughts there?

Yes; I think the issue is just that the "polygon" icon in Shapes is just a triangle, which is not really very polygonal. Reminds me of this great line in The Humans by Matt Haig, about an alien observing humans, commenting on our architecture: "The lack of geometric imagination was startling. There was not so much as a decagon in sight."[^1] 😂 So I think the lasso icon makes more sense for a polygon — then for lasso we could have a lasso that has vertices. What do you think?

> Maybe the completion_radius should be exposed in the Preferences? Then people could tweak based on monitor Rez and how intricate their labels are to be?

Yes, that was always the idea. 👍 

> Maybe it should accept None or something (infinity!) then the double-click always closes, just like the Shapes tools?

Yes. We could have the setting be a log slider, that starts at one (for the first-person-shooter crowd) maxes out at infinity. Also, the behaviour should be identical between the two layers and the preference should apply to both.

> Should we reconsider what the default behavior should be?

Yes, and it should match between the two layers. I like that it should be close but probably the default setting is too close.

> Maybe the Shapes tools should behave like the Labels tool does? But I'm loath to change something people are likely using and used to (plus my survey of tools I have access to suggests it's the more common behavior.)
Personally I think all 3 Polygon tools should have the same close behavior by default for consistency. And if there is a completion_radius setting it should affect all of them.

Yes, I agree with all of this. If we change it we should change it with a big release, but only when it's easy for users to set a preference that gets their existing behaviour back. But the two layers should absolutely be harmonised and indeed that should be the case for all the drawing tools, as we discussed when the Polygon tool for Labels was being added. We should have vector data and raster data and the same basic editing tools for both.

[^1]: later: "They placed me inside a small room that was, in perfect accord with all human rooms, a shrine to the rectangle." I do recommend the book. 😂

**Comment by psobolewskiPhD** [comment]
Discussed this at the community meeting today
https://hackmd.io/BXWDZ3i8Q6OAEASrkaSNIQ#2024-03-13

Ideally both shapes and labels annotations have tools that behave the same way and use\ an intermediate overlay before realizing either the vector shape or raster label. The later would also make it easier to implement multiscale `painting` as the overlay shape could be rasterized on each multiscale. After all, drawing rectangles or ellipses as labels can be useful too!

I agree the Shapes polygon icon (triangle) could be changed to a more polygon-y icon. I do like the current lasso icon for the functionality...

In the short term, I would use the Shapes polygon tool for the current labels tool functionality, with the change to match default behavior.
Then, the shapes lasso could be ported over to labels, allowing quick and easy auto-filling by just drawing contours. Here potentially spline interpolation would be nice? as noted by @brisvag in the original lasso tool PR.

**Comment by brisvag** [comment]
Ideally in my mind these 3 tools (4 if we imagine having both behaviours for labels as well) should all share the same code paths, with the only difference being the conditions for when a new vertex is added and when the shape is closed (mouse move vs mouse click/double click).

I would even suggest having a *single* icon/button for the two, and change behaviour depending on some simple interactions (i.e: double click to start or click-and-drag enters the "lasso" mode, while single click creates a normal vertex).

**Comment by psobolewskiPhD** [comment]
Just for reference: iN QuPath (I support a lot of QuPAth users at JAX) there are two tools (plus, rectangle, ellipse, brush, etc.):
![image](https://github.com/napari/napari/assets/76622105/2c1ee554-e818-427a-95d1-105c0b3522eb)

The left one `Polygon` does vertex by vertex polygons if you click to start, then double-click to finish -- they are always closed. This is identical to the current napari Shapes polygon tool (triangle icon).
If you drag, then you get a free-form polygon that will finish when you release drag -- again always closed. This is identical to the behavior of the Shapes lasso tool, with the exception that you don't need to hold button down, can just click and drag. 

The right QuPath button, Polyline, does vertex by vertex lines if you click to start, then double-click to finish -- they are always open. This is identical to the behavior of the napari `Path` tool.
However, the QuPath Polyline tool also supports free-form annotation if you just drag, that will finish when you release drag -- remains open. SO this is like the napari Shapes Polygon lasso, except it open. 
BTW, this free-form line tool is *exceptionally* useful for pixel annotation for training a classifier, because painting areas tends to generate too many similar samples. Would be worth adding!

So they have the same interactions as our shapes layer, but on 1 tool/icon for open and one for closed, while the vertex vs. freeform behavior gated by the user interaction (clicks vs. drag) essentially as @brisvag was suggesting.

**Comment by jni** [comment]
> So they have the same interactions as our shapes layer, but on 1 tool/icon for open and one for closed, while the vertex vs. freeform behavior gated by the user interaction (clicks vs. drag) 

Note that we created the single-click lasso for important ergonomic reasons, so I don't want this to be the only distinction. If we are running out of icon space (I'm sceptical but open), I suggest doing those little icons-with-a-down arrow, where a right-click or click-and-hold on the icon would let you choose the exact tool you want.

> Edit: Note that these tools create "Annotations" which are editable (via vertexes or the brush tool). QuPath can handle up to prolly 1K of those

I find this limitation weird, and probably due to an implementation detail rather than something fundamental? Obviously I'm just speculating here...

**Comment by brisvag** [comment]
Actually, of course we should unify the code not just for the 4 "closed" lasso tools, but also the open `Path` in shapes. Those work all exactly the same way in practice.

> Note that we created the single-click lasso for important ergonomic reasons, so I don't want this to be the only distinction.

I think a simple way to have all of these together is to have:

- single click followed by individual clicks makes vertices for a polygon. It can be terminated by either double-clicking, or clicking close enough to the starting vertex
- double click followed by mouse move makes a lasso. Terminate by clicking.
- click-and-drag makes a lasso. Terminate by mouse release.

**Comment by fjorka** [comment]
I really like the idea of a single icon and all different options as described by @brisvag!
There is one more difference between the lasso tool in the labels layer and the polygon in the shapes layer. The one in the labels layer responds to the right mouse click by removing the last added vertex - it would be great to keep this functionality for both labels and shapes.

**Comment by psobolewskiPhD** [comment]
@isabela-pf Do you have any ideas for a new polygon icon? How hard is it to make a SVG to add/replace
Specifically to replace the 5th in the 2nd row here, that looks like a triangle?
![image](https://github.com/napari/napari/assets/76622105/088c197e-7569-4beb-b239-50ead2f00bb5)
Here's the file: https://github.com/napari/napari/blob/main/napari/resources/icons/polygon.svg

The triangle doesn't feel very polygon-y 🤣 
I posted QuPath ones here https://github.com/napari/napari/issues/6489#issuecomment-1998329191

**Comment by psobolewskiPhD** [comment]
I've made a PR to address the icons
https://github.com/napari/napari/pull/7019
Here's the new look of the Shapes toolbar:
![image](https://github.com/napari/napari/assets/76622105/c4a4abdc-c0c4-4833-88ae-148da5fd8473)
The same icon is then used for the Labels polygon tool since it shares behavior with the Shapes one (ending behavior ignored for a moment).

**Comment by psobolewskiPhD** [comment]
Being able to change the minimum distance for completion (or disable it) has revealed that the last point is missing. E.g. it's not possible to make a triangle?

**Comment by psobolewskiPhD** [comment]
While working on the docs I

**Comment by jni** [comment]
Copying my response from the docs update:

> One last thing: in working on this PR i noticed that for Shapes, Esc completes, while for Labels Esc cancels and Enter completes. I dunno if we should fix that?

They should certainly be consistent.

> Escape to complete is a poor choice IMO.

Although I mostly agree, (a) Enter is awkward to reach if you are using a mouse right-handed, and (b) I just tried out Keynote and Esc completes. So whatever we choose should aim to be consistent with broader software out there. I don't have Photoshop & co to try out. (A lot of early Shapes work was done by @sofroniewn who is very familiar with Illustrator & co, so I *suspect* Esc to complete is actually reasonably standard.)

The other thing is that cancel is not that important an operation *as long as there is undo*, or easy deletion. I think at least one of these is true in some form or another for both labels and shapes. But of course we need to improve the discoverability of these things.

**Comment by psobolewskiPhD** [comment]
Yeah I agree. My concern is that Shapes has been around for a while so people may be using Esc already, so I'm wary of changing, thought I strongly believe double-click is much more natural.
I think with Shapes, if you hit Esc and get a shape, delete is really easy so no biggie to not have Esc cancel.
So maybe we just look into adding Enter to Shapes and leave Esc behavior?

**Comment by psobolewskiPhD** [comment]
Good news:
the Shapes behavior can be modified by the user:
<img width="776" alt="image" src="https://github.com/napari/napari/assets/76622105/07215445-42a8-4acc-a763-be910ec94877">
And so can the Labels:
<img width="630" alt="image" src="https://github.com/napari/napari/assets/76622105/0f150784-2e21-4b66-aa0e-410542ccc6fa">

So maybe this is a nothing burger?

**Comment by jni** [comment]
nah, sane defaults are useful, so it's at worst a broccoli burger. 😂

**Comment by psobolewskiPhD** [comment]
Ok, so are you suggesting just put `Enter` as the 2nd keybind for Shapes?

**Comment by psobolewskiPhD** [comment]
I'm going to pull out some outstanding elements:
- Shapes
   - Polygon: no way to remove a vertex like labels (right click)
   - Polygon: no abort like labels
- Labels
   - Autofill for free-hand annotations, see also: https://github.com/napari/napari/pull/5006 either via brush option or new lasso tool

Optional:
- merging button behaviors with either fly-out bars, right clicks, or mouse-click behaviors
